### PR TITLE
[1.x] Clean up test output for PhpStorm.

### PR DESCRIPTION
**Pull request by christophrumpel**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Fixed tickets | no ticket

This PR is a WIP to improve the PhpStorm output when running PEST through PhpStorm (not console).

**What needs to be improved:**
Here is what the current output looks like, which is very PHPUnit-like but not PEST-like.

CleanShot 2022-08-24 at 11 48 36@2x

Let's also compare it with the current PEST output in the console:

CleanShot 2022-08-24 at 11 58 26@2x

As you can see the PEST output is much cleaner and we also need this for PhpStorm output.

**What is included in this PR:**

By adapting the `Logging\TeamCity` class, I was able to reduce of the noise. Here is what the output looks like with my changes:

CleanShot 2022-08-24 at 11 46 44@2x

**Two issues still remain:**

1. The changes you can do in the `Logging\TeamCity` class are limited. This is also why @nunomaduro created the whole PEST output himself through PHPUnit events. (so it seems) Maybe it is better to try to use the same approach for this output? Is that possible?
2. Different to the console, PhpStorm stops at the first line of the output. That's why you always need to scroll to see the result, which you should see immediately.

CleanShot 2022-08-24 at 11 47 22@2x

Of course, it is even worse with the current output.

CleanShot 2022-08-24 at 11 48 51@2x

I think we have two option here: Try to see if we can scroll with PhpStorm plugin to the end of the output, or change the output to see the result first. We definitely want to see something green immediately to know the tests pass.

Thanks for the help so far @olivernybroe and I would appreciate some feedback and help here.

### CODE REVIEW

**Comment by olivernybroe** [comment]
I have create this issue to tackle the phpstorm side of things https://github.com/pestphp/pest-intellij/issues/364.

I think we should see if we can solve the scrolling in the plugin instead 👍

File: src/Logging/TeamCity.php [hunk 1]
```diff
@@ -50,20 +50,26 @@ public function __construct($out, bool $verbose, string $colors)
...
+     *
+     * @param string $buffer
+     * @return void
+     */
+    public function write(string $buffer): void
     {
-        $this->writeNewLine();
-        $this->write('Pest ' . version());
-        $this->writeNewLine();
+        if(str_contains($buffer, 'PHPUnit')) {
```
[the remark this answers is not in the document]
Comment (christophrumpel) [reply src/Logging/TeamCity.php:63]: Ok then let's try this. 👍

[the remark this answers is not in the document]
Comment (christophrumpel) [reply src/Logging/TeamCity.php]: As discussed, I now used a similar approach to the console output from the `PHPUnit Printer` class.

The output now looks very PEST-like ❤️:
CleanShot 2022-08-25 at 09 43 30@2x

I tested everything in PhpStorm manually, and it seems to be working. But please, @olivernybroe, take a look yourself. I deleted quite some code that I thought we didn't need anymore. But without tests, this is a little difficult to tell.

Do you think we should try to write some tests now too? Is there a way to run a whole test and check the output? I see that the `tests/Visual/output.txt` does not contain the full output.

[the remark this answers is not in the document]
Comment (olivernybroe) [reply src/Logging/TeamCity.php]: Amazing! I'll test it out locally :) 

Hmm, we could possibly just do snapshot testings?

**Comment by olivernybroe** [comment]
image
It looks like the stacktrace is not as clean anymore and that we actually have it twice, once in collision format and once in phpunit format 👀 
Here is a screenshot of the old one
image

Also it looks like the reported time to phpstorm and the time at the end does not match anymore. Not a big deal, but might be a bit confusing 🤔 
image


flowid is not filled out anymore, but I don't think we actually need the flow id before we do parallel stuff, so that can be handled at that point.


One thing to be aware of when touching the failure output is when it is a comparison. 
We do a special teamcity type called `comparisonFailure`, which adds this button
image that opens up to the following window
image


In general I think the improvements are really great! we just need to solve the last issues here with the stacktrace.

**Comment by fabio-ivona** [comment]
@christophrumpel test workflow is fixed now in 1.x branch

**Comment by christophrumpel** [comment]
@olivernybroe I did rewrite now most of the Printer to fix a few problems. I did this by storing all event data (like when a test failed, skipped, etc.) in an array and handling all the output in the `printResult` method together.

**This way, I was able to:**

* write a summary at the top as PEST does
* but still write the correct test output for single tests (which is a difference from PhpStorm to PEST)

This is what the top summary looks like:

CleanShot 2022-08-28 at 20 52 47@2x


**What we need to fix:**

I used your code to add the better "comparison output", but we this now two nice. (old and new way)
We need to get rid of the first (ugly) version.

CleanShot 2022-08-28 at 21 53 17@2x

Skipped and ignored tests look like this. It is the current PHPUnit way to do this. Not so pretty, but maybe something for later to fix.

CleanShot 2022-08-28 at 21 53 33@2x

Somehow the test name now looks like this, and I don't know why:

CleanShot 2022-08-28 at 22 07 35@2x

The time now also works, the show times are the same.

It would be great if you could help with the remaining things and test this version. I deleted quite some code too :-)

**Comment by olivernybroe** [comment]
The last one with the weird test name is because it is showing the class name of the test. When it starts with `P\` then it's an automatically generated class by pest.
To get the file name, there is a static method on the class we can call called `__getFileName`.


I found one other issue, running the tests on multiple files, groups them all under the same file. It seems like the testsuite started might not be triggered per testsuite (e.g. file).

The duplicated comparison output will be a bit tougher, it looks like that is handled by the IDE. I will see if I can get rid of the Expected/Actual text and just keep the link only 👍

**Comment by christophrumpel** [comment]
@olivernybroe I guess t we can close this one since you are have worked out a new way, right?

**Comment by olivernybroe** [comment]
@christophrumpel yes :) The new design will be very minimalistic instead and tailored to the IDE. 
It's a complete redesign

**Pull request closed without merging**